databases/core.py

```diff
@@ -63,8 +62,8 @@ def __init__(
         assert issubclass(backend_cls, DatabaseBackend)
         self._backend = backend_cls(self.url, **self.options)
 
-        # Connections are stored as task-local state.
-        self._connection_context: ContextVar = ContextVar("connection_context")
+        # Connections are stored per asyncio task
+        self._connections: typing.Dict[asyncio.Task, Connection] = {}
```

Comment:
Won't this leak memory? Is it maybe better to use a WeakKeyDictionary such that when the task is complete the key would be removed from the dictionary?

Reply:
The connection is dropped from the dict when it is closed — it shouldn't really leak memory.

Reply:
Two things here:
1. If tasks raised something and they weren't using context managers to safely close resources (eg: `transaction = database.transaction(); transaction.start(); raise ValueError` is enough) then cleanup code wouldn't run, so yes this could have leaked memory. 
2. Storing the `Task` as a workaround for avoiding asyncio's context variables prevented child tasks from inheriting active connections and transactions. I've gone back to `ContextVar`s, but the potential for strong references of a dictionary keeping forgotten / failed transactions and connections in memory still exists. I switched to `WeakKeyDictionary` like you suggested and added some tests to confirm the freeing of resources that didn't call teardown code.